transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java

```diff
@@ -1118,6 +1142,20 @@ private void callHandlerCallbackLater(AbstractChannelHandlerContext ctx, boolean
         }
     }
 
+    private void wakeUpForAutoFlush() {
+        if (autoFlush && ENQUEUE_WAKEUP_TASK_UPDATER.compareAndSet(this, 0, 1)) {
+            if (null == wakeupTask) {
+                wakeupTask = new Runnable() {
+                    @Override
+                    public void run() {
+                        ENQUEUE_WAKEUP_TASK_UPDATER.set(DefaultChannelPipeline.this, 0);
+                    }
+                };
+            }
+            channel().eventLoop().execute(wakeupTask);
+        }
+    }
+
```

Comment:
So, here is what my thought was: 

If I add these in the `AbstractChannelHandlerContext` then it gets invoked for every handler. When the code to refresh the cache `autoFlush` value is added, then that would mean touching a `volatile` field for every `write` on every handler. This code is an optimization to `wakeup` the selector on the first write (before the auto-flush task executes). Is it common that people call `write` on `ChannelHandlerContext` from outside the eventloop? Is it ok to add this _perceived_(as I don't have a benchmark) overhead on every handler invocation?